<a href="https://colab.research.google.com/github/imdann06/estructra_de_bases_de_datos/blob/main/Ejercicio_arboles.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Tienes un archivo con 10,000 estudiantes (nombre, edad,
promedio). Necesitas implementar un sistema que
permita:
1. Buscar un estudiante por su ID (número de matrícula)
2. Insertar nuevos estudiantes
3. Listar todos los estudiantes en orden por ID


In [99]:
pip install Faker

In [136]:
import random
from faker import Faker

def generar_estudiantes():
    fake = Faker()
    estudiantes = []

    for i in range(1, 10001):
        estudiante = {
            "id": i,
            "nombre": fake.first_name(),
            "promedio": round(random.uniform(0, 10), 1)
        }
        estudiantes.append(estudiante)

    return estudiantes


def buscar_estudiante(estudiantes):
    id_buscada = int(input("Ingrese el ID del estudiante a buscar: "))

    for estudiante in estudiantes:
        if estudiante["id"] == id_buscada:
            print(f'ID: {estudiante["id"]} | Nombre: {estudiante["nombre"]} | Promedio: {estudiante["promedio"]}')
            return

    print("Estudiante no encontrado")


def insertar_estudiante(estudiantes):
    id_est = len(estudiantes) + 1
    nombre = input("Ingrese nombre: ")
    promedio = float(input("Ingrese promedio: "))

    estudiantes.append({
        "id": id_est,
        "nombre": nombre,
        "promedio": promedio
    })


def listar_estudiantes(lista):
    for estudiante in lista:
        print(estudiante)


# generar los 10000 estudiantes
estudiantes = generar_estudiantes()


while True:

    print("\n--- MENU ---")
    print("1. Buscar estudiante")
    print("2. Insertar estudiante")
    print("3. Listar estudiantes")
    print("4. Salir")

    opcion = input("Seleccione una opción: ")

    if opcion == "1":
        buscar_estudiante(estudiantes)

    elif opcion == "2":
        insertar_estudiante(estudiantes)

    elif opcion == "3":
        listar_estudiantes(estudiantes)

    elif opcion == "4":
        print("Ha salido.")
        break

    else:
        print("Opción inválida")


--- MENU ---
1. Buscar estudiante
2. Insertar estudiante
3. Listar estudiantes
4. Salir
Seleccione una opción: 4
Ha salido.


In [174]:
#Tiempo de lista

import time

def medir_busquedas(estudiantes):

    ids_aleatorios = random.sample(range(1, len(estudiantes)+1), 100)

    inicio = time.time()

    for id_buscado in ids_aleatorios:

        for estudiante in estudiantes:
            if estudiante["id"] == id_buscado:
                break

    fin = time.time()

    print("Tiempo para 100 búsquedas:", fin - inicio, "segundos")


In [177]:
#100 estudiantes Lista
medir_busquedas(estudiantes)

Tiempo para 100 búsquedas: 0.10222601890563965 segundos


In [178]:
#construccion arbol ABB
class Nodo:
    def __init__(self, estudiante):
        self.estudiante = estudiante
        self.izquierda = None
        self.derecha = None


def insertar(nodo, estudiante):

    if nodo is None:
        return Nodo(estudiante)

    if estudiante["id"] < nodo.estudiante["id"]:
        nodo.izquierda = insertar(nodo.izquierda, estudiante)

    else:
        nodo.derecha = insertar(nodo.derecha, estudiante)

    return nodo


def buscar_abb(nodo, id_buscado):

    if nodo is None:
        return None

    if nodo.estudiante["id"] == id_buscado:
        return nodo.estudiante

    elif id_buscado < nodo.estudiante["id"]:
        return buscar_abb(nodo.izquierda, id_buscado)

    else:
        return buscar_abb(nodo.derecha, id_buscado)


# construir el ABB
raiz = None

for estudiante in estudiantes:
    raiz = insertar(raiz, estudiante)


import time
import random

def medir_busquedas_abb(raiz):

    ids = random.sample(range(1,10001),100)

    inicio = time.time()

    for i in ids:
        buscar_abb(raiz, i)

    fin = time.time()

    print("Tiempo ABB:", fin - inicio, "segundos")




In [180]:
#100 estudiantes ABB
medir_busquedas_abb(raiz)

Tiempo ABB: 0.0008003711700439453 segundos


In [106]:
pip install bplustree

In [181]:
from bplustree import BPlusTree
import random
import time

In [182]:
from bplustree import BPlusTree

arbol_bplus = BPlusTree("/tmp/estudiantes.db", order=50)

In [183]:
import os

if os.path.exists("/tmp/estudiantes.db"):
    os.remove("/tmp/estudiantes.db")

arbol_bplus = BPlusTree("/tmp/estudiantes.db", order=50)

In [184]:
import random
import time

def medir_busquedas_bplus(arbol):

    ids = random.sample(range(1,10001),100)

    inicio = time.time()

    for i in ids:
        arbol.get(i)

    fin = time.time()

    print("Tiempo B+:", fin - inicio, "segundos")

In [185]:
medir_busquedas_bplus(arbol_bplus)

Tiempo B+: 0.0020246505737304688 segundos


In [157]:
random.shuffle(estudiantes)

#COMPARACION DE TIEMPOS IDS ORDENADAS VS IDS EN DESORDEN PARA 100 BUSQUEDAS


ID ORDENADA

Lista: Tiempo para 100 búsquedas: 0.028342723846435547 segundos

ABB: Tiempo ABB:0.24692916870117188 segundos

B+: : Tiempo B+: 0.002257108688354492 segundos

ID DESORDENADA:

Lista: Tiempo para 100 búsquedas: 0.10222601890563965 segundos

ABB: Tiempo ABB: 0.0008308887481689453 segundos

B+: Tiempo B+: 0.0020246505737304688 segundos